tensorflow/core/graph/graph_constructor.cc

```diff
@@ -963,11 +966,17 @@ Status GraphConstructor::Convert() {
         // Note that input_already_exists can grow here
         AddControlDependencies(&imported_node_def, &input_already_exists);
       }
+      if (!opts_.bind_device.empty()) {
+        imported_node_def.set_device(opts_.bind_device);
```

Comment:
Should this be:

```c++
if (!opts_.default_device.empty() && imported_node_def.device().empty())
```

so we don't overwrite explicitly placed nodes? Nodes are often explicitly placed for a reason (for example, the user intended to force some computation to execute on CPU), and overriding that intent doesn't seem great.

By setting the device only for nodes that aren't explicitly placed, the behavior will be parallel to what happens in Python. For example:

```python
import tensorflow as tf

with tf.Graph().as_default():
  tf.constant(1.0, name='default')
  with tf.device('/gpu:0'):
    tf.constant(1.0, name='gpu')
  with tf.device('/cpu:0'):
    tf.constant(1.0, name='cpu')
  graph_def = tf.get_default_graph().as_graph_def()

# At this point graph_def has 3 nodes:
# one is explicitly placed on CPU, one on GPU, and one has no explicit placement.

with tf.Graph().as_default():
  with tf.device("/gpu:1"):
    tf.import_graph_def(graph_def)
  # This newly imported graph will have 3 nodes, keeping the explicit placements from before
  # and overriding the one node that wasn't placed
  print(tf.get_default_graph.as_graph_def())
```